src/Microsoft.AspNet.OData.Shared/NavigationPath.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using Microsoft.OData.UriParser;
+
+namespace Microsoft.AspNet.OData
+{
+    /// <summary>
+    /// Navigation Path of an OData ID
+    /// </summary>
+    public class NavigationPath
+    {
+        private string navigationPathName;
+        private ReadOnlyCollection<ODataPathSegment> _pathSegments;
+        private ConcurrentDictionary<string, PathItem[]> _pathItemCache = new ConcurrentDictionary<string, PathItem[]>();
+
+        /// <summary>
+        /// Constructor which takes and odataId and creates PathItems
+        /// </summary>
+        public NavigationPath()
+        {
+           
+        }
+
+        /// <summary>
+        /// Constructor which takes and odataId and creates PathItems
+        /// </summary>
+        /// <param name="navigationPath">ODataId in string format</param>
+        /// <param name="pathSegments">Pathsegment collection</param>
+        public NavigationPath(string navigationPath, ReadOnlyCollection<ODataPathSegment> pathSegments)
+        {
+            navigationPathName = navigationPath;
+            _pathSegments = pathSegments;           
+        }
+
+        
+        /// <summary>
+        /// NavigationPath/ODataId in string
+        /// </summary>
+        public string NavigationPathName { get { return navigationPathName; } }
+
+        /// <summary>
+        /// To Get ODataId in Parsed format
+        /// </summary>
+        /// <returns>Array of PathItems</returns>
+        public PathItem[] GetNavigationPathItems()
+        {
+            PathItem[] pathItems;
+            if(!_pathItemCache.TryGetValue(navigationPathName, out pathItems))
+            {
+                if (_pathSegments != null)
+                {
+                    pathItems = ParseODataId();
+                    _pathItemCache.TryAdd(navigationPathName, pathItems);
+                }
+            }
+
+            return pathItems;
+        }
```

Comment:
This method looks werid. 

It looks to retrieve a PathItem[] using a path name, but, the path name is private field? It's not from "outside".  Straigh?

In this case, why do we need a ConcurrencyDictionary?

Reply:
This was introduced to handle multiple set of path items in the same path like customers(1)/Orders(1)/shipment(1) and customers(1)/Orders(1)/Address(1). so cache customes(1)/orders(1) not to call again . But since we are making seperate navigation paths for each object  it wouldnt be ideal to keep a cache. Might be in later versions we can consider a centralizzed caching if we see such scenarios, even then we should think of heavy cache allocations . So removing it for now